### DISCUSSION TITLE
Mirth JSON to HL7 Error

### DISCUSSION THREAD

**Discussion by jgh1189** [nextgenhealthcare/connect#6212] [Q&A] [answered]
I'm receiving JSON files and need to convert to HL7. Currently working on converting the encounter info from the JSON file to an HL7 A08 message. Any help is appreciated. New to receiving data in this format!

I'm getting the following error message when testing a sample file:

Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException:
CHANNEL: Json_to_HL7
CONNECTOR: sourceConnector
SCRIPT SOURCE: TRANSFORMER
SOURCE CODE:
99: msg = JSON.parse(connectorMessage.getTransformedData());
100: function doFilter() { phase[0] = 'filter'; return true; }function doTransform() { phase[0] = 'transformer'; logger = Packages.org.apache.log4j.Logger.getLogger(phase[0]);
101:
102:
103: // Assuming your JSON data is read into the variable jsonData
104: var jsonData = JSON.parse(msg);
105:
106:
107: // Extract patient and visit data from JSON
108: var patient = jsonData.patientdetails;
LINE NUMBER: 104
DETAILS: SyntaxError: Unexpected token: o
at 68fd5082-ac4b-4128-9473-290f73409be7_JavaScript_Filter_Transformer_0:104 (doTransform)
at 68fd5082-ac4b-4128-9473-290f73409be7_JavaScript_Filter_Transformer_0:209 (doScript)
at 68fd5082-ac4b-4128-9473-290f73409be7_JavaScript_Filter_Transformer_0:211
at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:235)
at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:187)
at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:114)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
at java.base/java.lang.Thread.run(Thread.java:1570)

Source Transformer Code:
```
// Assuming your JSON data is read into the variable jsonData
var jsonData = JSON.parse(msg);

// Extract patient and visit data from JSON
var patient = jsonData.patientdetails;
var visit = jsonData.encountermetadata;

// Create HL7 segments
var MSH = [
"MSH",
"|",
"^~\&",
"SendingApp",
"SendingFac",
"ReceivingApp",
"ReceivingFac",
new Date().toISOString(),
"",
"ADT^A08",
"123456",
"P",
"2.3"
].join("|");

var EVN = [
"EVN",
"A08",
new Date().toISOString()
].join("|");

var PID = [
"PID",
"",
msg['patientdetails']['athenapatientid'],
"",
"",
msg['patientdetails']['lastname'] + "^" + msg['patientdetails']['firstname'],
"",
patient.dob,
patient.gender,
"",
patient.address
].join("|");

var PV1 = [
"PV1",
"",
"I",
"",
"",
"",
visit.attendingDoctor,
"",
"",
"",
"",
"",
"",
"",
"",
"",
"",
msg['encountermetadata'][0]['encounterid'],
"",
"",
"",
"",
"",
"",
"",
"",
"",
"",
"",
"",
msg['encountermetadata'][0]['encounterdate'],
""
].join("|");

// Combine segments into a single HL7 message
var hl7Message = [MSH, EVN, PID, PV1].join("\r");

// Set the transformed HL7 message as the outbound message
msg = hl7Message;
```

**Comment by kirbykn2** [comment]
What are your datatypes set to?
Also, test the incoming JSON to be sure its valid json.

**Reply by jgh1189** [reply]
Source: Inbound=JSON, Outbound = HL7 v.2x
Destination: Inbound=HL7 v.2x, Outbound = HL7 v.2x

I've since made some changes to my code and I'm getting an output HL7 message on the logger but I'm still getting an error:

"ER7Serializer error
ERROR MESSAGE: Error converting XML to ER7
org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog."

**Updated Code:**

```
// Check if msg is already a JSON object
var jsonData;
if (typeof msg === 'string') {
    // If it's a string, parse it to a JSON object
    jsonData = JSON.parse(msg);
} else {
    // If it's already an object, use it directly
    jsonData = msg;
}

// Ensure the structure is as expected
if (!jsonData || !jsonData.patientdetails || !Array.isArray(jsonData.encountermetadata)) {
    throw new Error("Invalid JSON structure");
}


var patient = jsonData.patientdetails;
var encounters = jsonData.encountermetadata;

function formatHL7Date(dateString) {
    var date = new Date(dateString);
    var year = date.getFullYear().toString().padStart(4, '0');
    var month = (date.getMonth() + 1).toString().padStart(2, '0');
    var day = date.getDate().toString().padStart(2, '0');
    var hours = date.getHours().toString().padStart(2, '0');
    var minutes = date.getMinutes().toString().padStart(2, '0');
    var seconds = date.getSeconds().toString().padStart(2, '0');
    return year + month + day + hours + minutes + seconds;
}

function createHL7Message(patient, encounter) {
    var MSH = [
        "MSH",
        "^~\\&",
        "ATHENA",
        "HF",
        "EPIC",
        "HF",
        formatHL7Date(new Date().toISOString()),
        "",
        "ADT^A08",
        "123456",
        "P",
        "2.3",
    ].join("|");

    var EVN = [
        "EVN",
        "",
        formatHL7Date(new Date().toISOString()),
        "",
        "",
        "",
        formatHL7Date(encounter.lastupdated || "")
    ].join("|");

    var PID = [
        "PID",
        "",
        "",
        patient.athenapatientid,
        "",
        patient.lastname + "^" + patient.firstname,
        "",
        formatHL7Date(patient.dob || "")
    ].join("|");

    var PV1 = [
        "PV1",
        "",
        "OP",
        encounter.departmentid,
        "",
        "",
        "",
        encounter.providerid,
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        encounter.encounterid,
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
	   "",
        "",
        formatHL7Date(encounter.encounterdate || ""),
        formatHL7Date(encounter.encounterdate || "")
    ].join("|");

    // Combine segments into a single HL7 message
    return [MSH, EVN, PID, PV1].join("\r");
}


// Create HL7 messages for each encounter
var hl7Messages = encounters.map(function(encounter) {
    return createHL7Message(patient, encounter);
});

logger.info("Parsed JSON Data: " + JSON.stringify(jsonData));
logger.info("Generated HL7 Messages: " + hl7Messages.join("\r"));

// Set the transformed HL7 messages as the outbound message
msg = hl7Messages.join("\r");

logger.info("Parsed JSON Data: " + JSON.stringify(jsonData));
logger.info("Generated HL7 Messages: " + hl7Messages.join("\r"));
```

**Reply by pacmano1** [reply]
**Format your code.**  **Use code blocks.**  Please fix your posts.   Read https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

This video is probably the tradtional way people do what you are asking: https://www.youtube.com/watch?v=pn_AlHfgqVI

**Reply by jgh1189** [reply]
Thanks for the feedback. Yes I have reviewed this video. However the JSON file that I'm receiving has multiple encounters within the file and the method in the video is only looking for 1. 

Back to my original solution, I feel that I am not far off. I'm still getting the error message: 

""ER7Serializer error
ERROR MESSAGE: Error converting XML to ER7
org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.""

Attaching the output of the transformation which appears to be transforming correctly. Is there something I'm missing for it to write out to the file? In my destination settings, should I change my template to something else--Currently have '${message.encodedData}'. Thanks. 
![Messages Tranformed output](https://github.com/nextgenhealthcare/connect/assets/109697053/cc0628d0-d78c-4535-829c-8f9553974d36)

![Logger details](https://github.com/nextgenhealthcare/connect/assets/109697053/d1893076-05bd-4b08-b368-a6d7fae04d71)

**Comment by kirbykn2** [comment]
I'm sorry, but wasn't this the error on the original post?  "DETAILS: SyntaxError: Unexpected token: o"

**Reply by jgh1189** [reply]
Correct. I updated my code. Sorry about that. If easier, I can close this and create a new discussion.

**Reply by kirbykn2** [reply]
yes, please put the original post back with the resolution.  Seems like this is a different discussion now.

**Comment by jgh1189** [comment, marked as answer]
This code addition resolved the original error: 'DETAILS: SyntaxError: Unexpected token: o'
```
// Check if msg is already a JSON object
var jsonData;
if (typeof msg === 'string') {
    // If it's a string, parse it to a JSON object
    jsonData = JSON.parse(msg);
} else {
    // If it's already an object, use it directly
    jsonData = msg;
}

// Ensure the structure is as expected
if (!jsonData || !jsonData.patientdetails || !Array.isArray(jsonData.encountermetadata)) {
    throw new Error("Invalid JSON structure");
}
```
